### PR TITLE
PYIC-3505: On-the-fly VC Status (4): Get status of last visited CRI VC on-the-fly in EvaluateGpg45ScoresHandle

### PR DESCRIPTION
**Pull request by TalatAydinGDS**
### What changed

Add new method (and tests for) getVCSuccessStatus to UserIdentityService. The new method will

- fetch the VCs from credential store
- return an optional indicating whether the VC for the given user and issuer is successful or not, if it exists
- return an empty optional otherwise

### Why did it change

EvaluateGpg45ScoresHandler determines the status of the last visited CRI VC on-the-fly

### Issue tracking

- [PYIC-3505](https://govukverify.atlassian.net/browse/PYIC-3505)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-3505]: https://govukverify.atlassian.net/browse/PYIC-3505?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java [hunk 1]
```diff
@@ -300,6 +272,9 @@ void shouldStoreRequiredScoresInSessionIfGpg45ProfileNotMatched() throws Excepti
 
         evaluateGpg45ScoresHandler.handleRequest(request, context);
 
+        ArgumentCaptor<IpvSessionItem> ipvSessionItemArgumentCaptor =
+                ArgumentCaptor.forClass(IpvSessionItem.class);
+        verify(ipvSessionService).updateIpvSession(ipvSessionItemArgumentCaptor.capture());
```
Comment (Wynndow) [inline lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: You don't need to capture the session like this - I recently merged a change that made the session object a spy (`@Spy private IpvSessionItem ipvSessionItem;`).

You're not actually doing anything with that captured argument here either, so you could just call 
```
verify(ipvSessionService).updateIpvSession(ipvSessionItem);
```

Are you looking to ensure that the session was persisted after the required scores were added? If so, you might want to look at what I did recently after getting bitten and causing a P1. Need to check that the update happens *after* the write.  You could change the below to work for writing the required scores.

```
        InOrder inOrder = inOrder(ipvSessionItem, ipvSessionService);
        inOrder.verify(ipvSessionItem).setVot(VOT_P2);
        inOrder.verify(ipvSessionService).updateIpvSession(ipvSessionItem);
        inOrder.verify(ipvSessionItem, never()).setVot(any());
```

File: lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java [hunk 2]
```diff
@@ -542,12 +486,6 @@ void shouldSendAuditEventWhenProfileMatched() throws Exception {
                 List.of("123ab93d-3a43-46ef-a2c1-3c6444206408", "RB000103490087", "abc1234"),
                 extension.getVcTxnIds());
         verify(clientOAuthSessionDetailsService, times(1)).getClientOAuthSession(any());
-
-        InOrder inOrder = inOrder(ipvSessionItem, ipvSessionService);
-        inOrder.verify(ipvSessionItem).setVot(VOT_P2);
-        inOrder.verify(ipvSessionService).updateIpvSession(ipvSessionItem);
-        inOrder.verify(ipvSessionItem, never()).setVot(any());
-        assertEquals(VOT_P2, ipvSessionItem.getVot());
```
Comment (Wynndow) [inline lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: Need to keep this in - quite important.

File: lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java [hunk 3]
```diff
@@ -630,6 +630,7 @@ void shouldReturnJourneyEvaluateResponseOnSuccessfulPendingCriResponse() {
                 .thenReturn(
                         VerifiableCredentialResponse.builder()
                                 .userId(TEST_USER_ID)
+                                .verifiableCredentials(List.of(SignedJWT.parse(SIGNED_VC_1)))
```
Comment (Wynndow) [inline lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java]: I don't think you should need to add this - a `pending` response wouldn't have a VC attached to it.

**Review by Wynndow — COMMENTED** [review]
Couple of small comments about tests. I've deployed this to my dev env too and am currently running the feature tests against it.

**Comment by Wynndow** [comment]
~The feature tests are showing an issue with this PR: https://reports.cucumber.io/reports/76a3f570-42be-437a-b993-fe39f71c721c~

~Looks like a user can run a successful identity proving journey but is issued a P0. This is similar to issues we've seen recently. Will take a closer look tomorrow.~

Scratch that - I might have deployed on older version of your PR. Will re-deploy and run the tests again

**Comment by Wynndow** [comment]
Yeah feature tests are actually fine.

File: lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java (hunk 1 above)
Comment (TalatAydinGDS) [reply lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: Thanks comments. Done.

File: lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java (hunk 2 above)
Comment (TalatAydinGDS) [reply lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: Done.

File: lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java (hunk 3 above)
Comment (TalatAydinGDS) [reply lambdas/retrieve-cri-credential/src/test/java/uk/gov/di/ipv/core/retrievecricredential/RetrieveCriCredentialHandlerTest.java]: Done.

File: lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java [hunk 4]
```diff
@@ -308,6 +281,9 @@ M1A, new Gpg45ScoresDto(List.of(new EvidenceDto(4, 2)), 0, 0, 2)),
                                 M1B, new Gpg45ScoresDto(List.of(new EvidenceDto(3, 2)), 1, 0, 2))),
                 ipvSessionItem.getRequiredGpg45Scores());
 
+        InOrder inOrder = inOrder(ipvSessionItem, ipvSessionService);
+        inOrder.verify(ipvSessionService).updateIpvSession(ipvSessionItem);
+        inOrder.verify(ipvSessionItem, never()).setVot(any());
```
Comment (Wynndow) [inline lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: I don't think you need this here. My original comment was around if you wanted to check that the required scores had been stored in the session (that's what this test is testing). We're already checking that the VOT isn't set in the lines just below.

If you're not worried about testing that the required scores have been written correctly I probably wouldn't change this test at all. Considering what your PR is for, you probably don't need to be updating this test at all.

Comment (TalatAydinGDS) [reply lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: Done.

Comment (TalatAydinGDS) [reply lambdas/evaluate-gpg45-scores/src/test/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandlerTest.java]: When we ran the feature test together with test team, we encountered some errors, so we had to take the latest changes you made in the main branch and changes some unit tests. I have reverted this test to its previous state.

**Review by Wynndow — APPROVED** [review]

**Pull request merged**